SDL/SDLJoystick.cpp

```diff
@@ -11,167 +11,209 @@ extern "C" {
-			key.keyCode = NKCODE_DPAD_RIGHT;
-			NativeKey(key);
-#else
-			AxisInput axisX;
-			AxisInput axisY;
-			axisX.axisId = JOYSTICK_AXIS_HAT_X;
-			axisY.axisId = JOYSTICK_AXIS_HAT_Y;
-			axisX.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-			axisY.deviceId = DEVICE_ID_PAD_0 + deviceIndex;
-			axisX.value = 0.0f;
-			axisY.value = 0.0f;
-			if (event.jhat.value & SDL_HAT_LEFT) axisX.value = -1.0f;
-			if (event.jhat.value & SDL_HAT_RIGHT) axisX.value = 1.0f;
-			if (event.jhat.value & SDL_HAT_DOWN) axisY.value = 1.0f;
-			if (event.jhat.value & SDL_HAT_UP) axisY.value = -1.0f;
-			NativeAxis(axisX);
-			NativeAxis(axisY);
-#endif
-			break;
 		}
-
-	case SDL_JOYDEVICEADDED:
+		case SDL_CONTROLLERDEVICEREMOVED:
 		{
-			int deviceIndex = event.jdevice.which;
-			if (deviceIndex >= joys.size()) {
-				joys.resize(deviceIndex+1);
+			// for removal events, "which" is the instance ID for SDL_CONTROLLERDEVICEREMOVED
+			if (event.cdevice.which == player1JoyInstanceID)
+			{
+				printf("control pad removed for player 1\n");
+				if (player1Controller){
+						SDL_GameControllerClose(player1Controller);
+						player1Controller = NULL;
+				}
+				// are there any other control pads we could use?
+				setUpPlayer1();
 			}
-			joys[deviceIndex] = SDL_JoystickOpen(deviceIndex);
-			SDL_JoystickEventState(SDL_ENABLE);
 			break;
 		}
-
-	case SDL_JOYDEVICEREMOVED:
+		case SDL_CONTROLLERDEVICEADDED:
 		{
-			int deviceIndex = getDeviceIndex(event.jdevice.which);
-			if (deviceIndex >= 0) {
-				SDL_JoystickClose(joys[deviceIndex]);
-				joys[deviceIndex] = 0;
+			// for add events, "which" is the device index!
+			if (!player1Controller)
+			{
+				if (SDL_IsGameController(event.cdevice.which))
+				{
+					player1Controller = SDL_GameControllerOpen(event.cdevice.which);
+					SDL_GameControllerEventState(SDL_ENABLE);
+					if (player1Controller && SDL_GameControllerGetAttached(player1Controller))
+					{
+						player1JoyInstanceID = SDL_JoystickInstanceID(SDL_GameControllerGetJoystick(player1Controller));
+						printf("Player 1 will be using control pad: %s, instance ID: %d\n", SDL_GameControllerName(player1Controller), player1JoyInstanceID);
+						char *mapping = SDL_GameControllerMapping(player1Controller);
+						if (mapping == NULL)
+						{
+							printf("control pad %s is not in the game controller database, unable to set mappings!\n", SDL_GameControllerName(player1Controller));
+						}
+						else
+						{
+							printf("control pad mapping: %s\n", SDL_GameControllerMapping(player1Controller));
+						}
+					}
+					else
+					{
+						printf("Error! Could not open controller\n");
+					}
+				}
 			}
 			break;
 		}
 	}
 }
 
 int SDLJoystick::getDeviceIndex(int instanceId) {
-	for (int i = 0; i < joys.size(); i++) {
-		SDL_Joystick *joy = joys[i];
-		if (SDL_JoystickInstanceID(joy) == instanceId) {
+	int numjoys = SDL_NumJoysticks();
+	SDL_Joystick *joy;
+	for (int i = 0; i < numjoys; i++)
+	{
+		joy = SDL_JoystickOpen(i);
```

Comment:
Hmm, is this fast?  It might be nice if we could keep around the instance ids for activated controllers in an array since we handle add/remove anyway.  I guess this probably isn't so slow as to add any significant latency to pad events...

-[Unknown]